### DISCUSSION TITLE
Questions about support and training of specific code languages

### DISCUSSION THREAD

**Discussion by MOomenAL** [community/community#86112] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

Hello Github,

As a developer I work with the AL code language which is specific for Microsoft Business Central. Due to this and other historical reasons there is not much code publicly available, especially if you compare it with some other languages. 
Since a few months a couple of colleagues and myself are using the Github Copilot to see if this is a tool usable for our wider department, and as we expected due to the reasons above the suggestions are hit and miss. But we were hoping that by working with Copilot the output would improve, since we were providing input to the model. 
However after some more research of some of the documentation I understand that is not the case? That model is trained only by available public code, not by user input/prompts? Is that correct?
If that is the case, how would handling of lesser used code languages by Copilot improve? Next to the obvious of increasing the publicly available code.
Thanks in advance for your replies.

**Comment by Akash1134** [comment, marked as answer]
Hi @MOomenAL ,

GitHub Copilot is context aware of the code and language you use in your editor. We've been making instrumental progress in increasing Copilot context awareness. You can read more about it [here](https://github.blog/2023-05-17-how-github-copilot-is-getting-better-at-understanding-your-code/). 

In general, GitHub Copilot is trained on all languages that appear in public repositories. For each language, the quality of suggestions you receive may depend on the volume and diversity of training data for that language. For example, JavaScript is well-represented in public repositories and is one of GitHub Copilot's best supported languages. Languages with less representation in public repositories may produce fewer or less robust suggestions. 

Broadly, there are two ways to improve suggestions for any particular language - one is to increase the usability which will improve with time as more users will a specific language and the training models have a wide variety of code to offer suggestions from. Another way is increasing the context awareness which I think is playing a larger role. One technique we use today is to look for relevant code in neighbouring files (i.e. tabs open in the code editor).
Imagine that you have two files open – one that defines a Class, and another with unit tests for that Class. By looking at both files, we have more context into the relevant code, which results in higher-quality suggestions.

**Reply by MOomenAL** [reply]
Hi @Akash1134,

Thank you for your reply. I do have a follow up question; I understand how to use the suggestions for improvement for myself, but I wonder if, and if so, how that has consequences for my fellow AL developers? That example you mentioned would lead to better suggestions in my case, but would my acceptance of a suggestion also lead to a higher quality suggestion down the road for my colleagues working on similar problem on their machine?
Addendum; Would my example lead to better suggestions for myself in a similar situation 2 weeks later?

**Reply by Akash1134** [reply]
>  I understand how to use the suggestions for improvement for myself, but I wonder if, and if so, how that has consequences for my fellow AL developers? 

Yes, that goes both ways. As you and other AL developers continue to create wider contexts for Copilot, LLMs will get better and hence Copilot will have improved output in the future. Note, that this has to be done on a publically available code. See: [Copilot FAQs](https://github.com/features/copilot)

**Reply by MOomenAL** [reply]
Clear, thank you.
